[Environment] Infra parts (#43)
SOMAS2020/SOMAS-Final-Report: 04_environment/infra.tex

\chapter{Environment Infrastructure}
\section{Foraging}
\subsection{Inputs, Outputs and Functions}

> Foraging has been implemented in the code base. It determines the method of foraging. In short, it is either deer hunting being payoff focused or fishing being a risk averse method of foraging. Deer hunting is linked to a population model, based on a logistic population model. The high level inputs include island IDs, resource contributions from those islands and the chosen foraging method. The high level output is a foraging report which include forage type, input resources, number of participants, number caught, total utility, catch sizes, turn.\newline

Reviewer: ```suggestion
Foraging has been implemented in the code base. It determines the method of foraging. In short, it is either deer hunting being payoff focused or fishing being a risk averse method of foraging. Deer hunting is linked to a logistic population model. When foraging starts, all islands are asked for their resource contributions and chosen foraging method. The high level output is a foraging report which include forage type, input resources, number of participants, number caught, total utility, catch sizes, turn.\newline
```

Did you mean logarithmic instead of logistics?

Reviewer 2: Hey, 
I was informed that it is logistics,

                                     & \texttt{\begin{tabular}[c]{@{}c@{}}CreateDeer\\ PopulationModel\end{tabular}} & \begin{tabular}[c]{@{}l@{}}Returns the target\\ population model.\end{tabular}                                                                                    & \begin{tabular}[c]{@{}l@{}}- Deer hunt \\ configuration\\ - Logger\end{tabular}                                                                             & - Population model                                                                                                                                                          \\ \hline
\end{tabular}
\end{center}
\caption{Foraging Main Functions}
\label{tab:Foraging Main Functions}
\end{table}

\newpage
\subsection{Important Implementation Remarks}

> \subsubsection{Utility for a Single Deer}

Reviewer: This is repeated in the design environment section. May be worth removing or just referencing the other section

Reviewer 2: Hey @NGovani 
Are you referring to the complete subsection or the subsubsection?
There are similar parts and the same equations as the design doc but we don't believe it should remove the entire subsection as for example deer population is described here.

Reviewer: The subsubsections for utility is what I meant. It feels like it describes a similar process to what happens in tiers in design section of the environment along with the Bernoulli, if you think it helps to combine all of them into a single part (considering they are spread out in the design section) then it should be fine to keep.

Reviewer 2: Hey, 
so after discussing a little bit more with the group, 
we do describe the same concepts as in the design doc but in wider details and depth and elaborate a lot more on the math. Therefore, they are similar but different. 
may we keep it as is and refer it to the design section?
the only prob is that the design doc has been already merged 
Let me know what you think : )

Reviewer 2: > The subsubsections for utility is what I meant. It feels like it describes a similar process to what happens in tiers in design section of the environment along with the Bernoulli, if you think it helps to combine all of them into a single part (considering they are spread out in the design section) then it should be fine to keep.

Great! Thanks, we will remove the utility section and a few equations that are the same as the design doc  :)

The return for a single deer is modelled as a random variable (RV) that is implemented as \texttt{deerReturn()}. It is effectively the combination of two other RVs:
\begin{itemize}
    \item $D$: Bernoulli RV represents the binary probability of catching a deer. Usually Bernoulli probability $p$ (i.e. $P(D=1) = p$), will be fairly close to $1$ such that there is a fairly high chance of catching a deer if agents invest sufficient resources.
    \item $W$: A continuous RV that adds some variance to the return. This could be interpreted as the weight/size of the deer that is caught. $W$ is exponentially distributed such that the prevalence of deer is inversely proportional to their size.
\end{itemize}

$D$ and $W$ are combined in the expression $H = D(1+W)$ which, in turn, is another RV which corresponds to the output of \texttt{deerReturn()}. Notice how $W$ is irrelevant if $D=0$; the weight of a deer does not matter if it evades capture. Also, the mean return can be shown to be $$E[U] = E[D(1+W)] = p(1+E[W]) = p(1+\frac{1}{\lambda})$$ where $p$ is the Bernoulli probability of $D$ and $\lambda$ is the exponential rate parameter of $W$.

> \subsubsection{Dynamics of a Collective Hunt}

Reviewer: this is a good expansion on the design part, and should definitely stay

The utility distribution is modelled as being agnostic to the number of hunters; it only depends on the collective amount of resources invested in the hunt. Furthermore, assume the minimum resources to hunt a single deer is $\theta$ and employ the assumption that hunting one deer is independent to another. If we let $\theta=1$\footnote{it is desirable to use 1 in this definition to allow for later scaling by an arbitrary constant to ensure a scale that is commensurate with other resource-based quantities in the environment} and define the expected return for a single deer as $E[U]=\mu$, for a given real-valued collective resource input $x$ (across all hunt participants), the expected total utility $U_{\theta}$ is proposed as follows:

\begin{equation}
U_{\theta}=\left\{\begin{array}{ll}
\mu & x \in[\Delta^0; \ \Delta^1)\\
2 \mu & x\in[\Delta^1; \ \Delta^2)\\
3 \mu & x\in[\Delta^2; \ \Delta^3)\\
4 \mu & x\in[\Delta^3; \ \Delta^4)\\
\end{array}\right.  \quad \text{where} \quad \Delta^k= \sum_{i=0}^k\delta^i  \quad 
\end{equation}

> Where $n=4$ (maximum number of deer that can be hunted in a single hunt). Notice that the expected return for $n$ deer is simply $n$ times the return of a single deer (i.i.d. assumption). However, the incremental input resources required to hunt $n$ deer decreases as $n$ increases. That is, it costs less (per deer) to hunt more deer - a dynamic that incentivises collaboration amongst agents. This is incorporated using a variable $\delta$ (\texttt{decay} $=$ \texttt{gameConf.ForagingConfig.IncrementalInputDecay}) which is used as follows: for each additional deer (above 1), the incremental cost required to hunt it is calculated by $\delta^n$. This decay parameter $\delta$ should be chosen such that $0< \delta< 1$. Assuming $\delta=0.8$ at the time of implementation 

Reviewer: Should be mentioned that this a parameter than can be changed

\caption{Main funcitions of Common Pool for Disaster Mitigation}
\label{tab: Common Pool's Main Function}
\end{table}

This section details the implementation of Common Pool disaster mitigation. This includes a predetermined, constant \texttt{threshold} value that effectively represents a minimum required level of collective `insurance` against disasters. If a disaster occurs and the Common Pool resource level exceeds this threshold, the effects of the disaster are halved. If not, no such halving takes place and the full effect of the disaster is felt by the islands.

An important input to the implementation is the magnitude of the disaster’s damage on each island. In Figure~\ref{fig:Common Pool infrastructure outcome}, this specific disaster’s magnitude was 3.458. However, the disaster’s effect on individual islands does not add up to 3.458 because the storm’s epicenter is located some distance away from each island. This aspect creates a more realistic scenario. From the individual damage experienced by each island, proportional effects on each island are computed. The summation of these proportional effects is equal to 100$\%$, which is the total effective damage.\newline

> If the common pool has enough resources to fully mitigate the disaster, the process is very easy as each island escapes unscathed. However, in the unfortunate event that the total disaster damage exceeds Common Pool resources, all such resources will be distributed to islands in proportion to the damage they experience (regardless of what their prior Common Pool contributions were). This is also visualised in Figure~\ref{fig:Common Pool infrastructure outcome}: Island 4 experiences the most damage from the storm and so, Island 4 receives the maximum Common Pool resource allocation of 252.47 resources (“common pool mitigated” column). The updated damage value shows the leftover damage of the disaster that the common pool was not able to mitigate. Left over damage serves to deplete an island's private pool of resources where applicable. If there is enough resources in the common pool to totally mitigate the disaster, the update damage values will be zero.

Reviewer: It should be explicitly stated that the disaster first takes away from the common pool and then affect the islands. 
```suggestion
When a disaster strikes the effect of the disaster on each islands is summed and then compared to the resources in the common pool. If the common pool has more resources than the total effect of the disaster it can fully mitigate the disaster meaning the islands take not damage. However, in the unfortunate event that the total disaster damage exceeds Common Pool resources, the common is first depleted and the total disaster effect is also reduced by the same amount and then the damage is applied to islands in proportion to the original damage they would have experienced (regardless of what their prior Common Pool contributions were). This is also visualised in Figure~\ref{fig:Common Pool infrastructure outcome}: Island 4 experiences the most damage from the storm and so, Island 4 receives the maximum Common Pool resource allocation of 252.47 resources (“common pool mitigated” column). The updated damage value shows the leftover damage of the disaster that the common pool was not able to mitigate. Left over damage serves to deplete an island's private pool of resources where applicable. If there is enough resources in the common pool to totally mitigate the disaster, the update damage values will be zero.
```

Apologies for the long suggestion

Reviewer 2: Thank you : )
